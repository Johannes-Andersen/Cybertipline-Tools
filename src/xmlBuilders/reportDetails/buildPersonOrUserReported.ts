import { XMLBuilder, XMLParser } from 'fast-xml-parser';
import type { PersonOrUserReported } from '../../types';
import { buildDeviceId } from '../common/buildDeviceId';
import { buildEstimatedLocation } from '../common/buildEstimatedLocation';
import { buildIpCapture } from '../common/buildIpCapture';
import { buildPerson } from '../common/buildPerson';

const parser = new XMLParser({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  allowBooleanAttributes: true,
});

const builder = new XMLBuilder({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  suppressBooleanAttributes: false,
});

export const buildPersonOrUserReported = (
  personOrUserReported: PersonOrUserReported,
  keyName = 'personOrUserReported',
): string => {
  const {
    estimatedLocation,
    personOrUserReportedPerson,
    deviceId,
    ipCaptureEvent,
  } = personOrUserReported;

  const deviceIds = deviceId?.map(
    (e) => parser.parse(buildDeviceId(e)).deviceId,
  );
  const ipCaptureEvents = ipCaptureEvent?.map(
    (e) => parser.parse(buildIpCapture(e)).ipCaptureEvent,
  );
  const estimatedLocations = estimatedLocation
    ? parser.parse(buildEstimatedLocation(estimatedLocation)).estimatedLocation
    : undefined;
  const person = personOrUserReportedPerson
    ? parser.parse(buildPerson(personOrUserReportedPerson)).person
    : undefined;

  return builder.build({
    [keyName]: {
      personOrUserReportedPerson: person,
      vehicleDescription: personOrUserReported.vehicleDescription,
      espIdentifier: personOrUserReported.espIdentifier,
      espService: personOrUserReported.espService,
      screenName: personOrUserReported.screenName,
      displayName: personOrUserReported.displayName,
      profileUrl: personOrUserReported.profileUrl,
      ipCaptureEvent: ipCaptureEvents,
      deviceId: deviceIds,
      thirdPartyUserReported: personOrUserReported.thirdPartyUserReported,
      priorCTReports: personOrUserReported.priorCTReports,
      groupIdentifier: personOrUserReported.groupIdentifier,
      estimatedLocation: estimatedLocations,
      additionalInfo: personOrUserReported.additionalInfo,
    },
  });
};
