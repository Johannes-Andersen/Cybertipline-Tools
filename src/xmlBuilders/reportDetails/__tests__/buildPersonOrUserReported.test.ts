import { describe, expect, it } from 'vitest';
import {
  Country,
  DeviceEventName,
  DeviceIdType,
  IPEventName,
  State,
} from '../../../types';
import type { PersonOrUserReported } from '../../../types';
import { buildPersonOrUserReported } from '../buildPersonOrUserReported';

describe('buildPersonOrUserReported', () => {
  it('should build a person or user reported with all fields', () => {
    const personOrUserReported: PersonOrUserReported = {
      personOrUserReportedPerson: {
        firstName: 'James',
        lastName: 'Wilson',
        age: 35,
        dateOfBirth: new Date('1989-03-15'),
        phone: [
          {
            number: '555-0123',
            verified: true,
          },
        ],
        email: [
          {
            email: 'jwilson@example.com',
            verified: true,
          },
        ],
        address: [
          {
            address: '123 Main St',
            city: 'Springfield',
            state: State.Connecticut,
            zipCode: '62701',
          },
        ],
      },
      vehicleDescription: 'Red Toyota Camry, License Plate ABC-123',
      espIdentifier: 'USER123456',
      espService: 'Social Media Platform X',
      screenName: 'jwilson_1989',
      displayName: ['James W.', 'JW'],
      profileUrl: ['https://platform-x.com/jwilson_1989'],
      ipCaptureEvent: [
        {
          ipAddress: '192.168.1.100',
          eventName: IPEventName.Login,
          dateTime: new Date('2024-01-15T10:30:00Z'),
          possibleProxy: false,
          port: 443,
        },
        {
          ipAddress: '10.0.0.50',
          eventName: IPEventName.Upload,
          dateTime: new Date('2024-01-15T11:45:00Z'),
          possibleProxy: true,
        },
      ],
      deviceId: [
        {
          idType: DeviceIdType.IMEI,
          idValue: '123456789012345',
          eventName: DeviceEventName.Login,
          dateTime: new Date('2024-01-15T10:30:00Z'),
        },
      ],
      thirdPartyUserReported: false,
      priorCTReports: [123456, 789012],
      groupIdentifier: 'Group XYZ',
      estimatedLocation: {
        city: 'Springfield',
        region: 'Illinois',
        countryCode: Country.UnitedStates,
        verified: true,
        timestamp: new Date('2024-01-15T12:00:00Z'),
      },
      additionalInfo:
        'Subject frequently changes usernames and uses multiple devices',
    };

    const result = buildPersonOrUserReported(personOrUserReported);

    expect(result).toMatchSnapshot();
  });

  it('should build a person or user reported with minimal online presence', () => {
    const personOrUserReported: PersonOrUserReported = {
      screenName: 'anonymous_user',
      ipCaptureEvent: [
        {
          ipAddress: '192.168.1.200',
          eventName: IPEventName.Upload,
        },
      ],
      thirdPartyUserReported: true,
    };

    const result = buildPersonOrUserReported(personOrUserReported);

    expect(result).toMatchSnapshot();
  });

  it('should build a person or user reported with only physical details', () => {
    const personOrUserReported: PersonOrUserReported = {
      personOrUserReportedPerson: {
        firstName: 'Robert',
        lastName: 'Brown',
        age: 42,
        address: [
          {
            address: '456 Oak Street',
            city: 'Chicago',
            state: State.Illinois,
          },
        ],
      },
      vehicleDescription: 'Blue Ford F-150',
      estimatedLocation: {
        city: 'Chicago',
        region: 'Illinois',
        countryCode: Country.UnitedStates,
      },
    };

    const result = buildPersonOrUserReported(personOrUserReported);

    expect(result).toMatchSnapshot();
  });

  it('should build a person or user reported with multiple devices and IPs', () => {
    const personOrUserReported: PersonOrUserReported = {
      screenName: 'tech_user',
      deviceId: [
        {
          idType: DeviceIdType.IMEI,
          idValue: '987654321098765',
          eventName: DeviceEventName.Login,
        },
        {
          idType: DeviceIdType.ICCID,
          idValue: '12345678901234567890',
          eventName: DeviceEventName.Registration,
        },
      ],
      ipCaptureEvent: [
        {
          ipAddress: '172.16.0.100',
          eventName: IPEventName.Login,
          possibleProxy: false,
        },
        {
          ipAddress: '172.16.0.200',
          eventName: IPEventName.Upload,
          possibleProxy: true,
        },
      ],
      priorCTReports: [111222],
      additionalInfo: 'Multiple devices used within short time spans',
    };

    const result = buildPersonOrUserReported(personOrUserReported);

    expect(result).toMatchSnapshot();
  });
});
