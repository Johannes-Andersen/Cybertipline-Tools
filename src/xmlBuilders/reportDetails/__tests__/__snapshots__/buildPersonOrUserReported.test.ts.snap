// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildPersonOrUserReported > should build a person or user reported with all fields 1`] = `"<personOrUserReported><personOrUserReportedPerson><firstName>James</firstName><lastName>Wilson</lastName><phone verified="true">555-0123</phone><email verified="true">jwilson@example.com</email><address><address>123 Main St</address><city>Springfield</city><zipCode>62701</zipCode><state>CT</state></address><age>35</age><dateOfBirth>1989-03-15</dateOfBirth></personOrUserReportedPerson><vehicleDescription>Red Toyota Camry, License Plate ABC-123</vehicleDescription><espIdentifier>USER123456</espIdentifier><espService>Social Media Platform X</espService><screenName>jwilson_1989</screenName><displayName>James W.</displayName><displayName>JW</displayName><profileUrl>https://platform-x.com/jwilson_1989</profileUrl><ipCaptureEvent><ipAddress>192.168.1.100</ipAddress><eventName>Login</eventName><dateTime>2024-01-15T10:30:00.000Z</dateTime><possibleProxy>false</possibleProxy><port>443</port></ipCaptureEvent><ipCaptureEvent><ipAddress>10.0.0.50</ipAddress><eventName>Upload</eventName><dateTime>2024-01-15T11:45:00.000Z</dateTime><possibleProxy>true</possibleProxy></ipCaptureEvent><deviceId><idType>IMEI</idType><idValue>123456789012345</idValue><eventName>Login</eventName><dateTime>2024-01-15T10:30:00.000Z</dateTime></deviceId><thirdPartyUserReported>false</thirdPartyUserReported><priorCTReports>123456</priorCTReports><priorCTReports>789012</priorCTReports><groupIdentifier>Group XYZ</groupIdentifier><estimatedLocation verified="true" timestamp="2024-01-15T12:00:00.000Z"><city>Springfield</city><region>Illinois</region><countryCode>US</countryCode></estimatedLocation><additionalInfo>Subject frequently changes usernames and uses multiple devices</additionalInfo></personOrUserReported>"`;

exports[`buildPersonOrUserReported > should build a person or user reported with minimal online presence 1`] = `"<personOrUserReported><screenName>anonymous_user</screenName><ipCaptureEvent><ipAddress>192.168.1.200</ipAddress><eventName>Upload</eventName></ipCaptureEvent><thirdPartyUserReported>true</thirdPartyUserReported></personOrUserReported>"`;

exports[`buildPersonOrUserReported > should build a person or user reported with multiple devices and IPs 1`] = `"<personOrUserReported><screenName>tech_user</screenName><ipCaptureEvent><ipAddress>172.16.0.100</ipAddress><eventName>Login</eventName><possibleProxy>false</possibleProxy></ipCaptureEvent><ipCaptureEvent><ipAddress>172.16.0.200</ipAddress><eventName>Upload</eventName><possibleProxy>true</possibleProxy></ipCaptureEvent><deviceId><idType>IMEI</idType><idValue>987654321098765</idValue><eventName>Login</eventName></deviceId><deviceId><idType>ICCID</idType><idValue>12345678901234567890</idValue><eventName>Registration</eventName></deviceId><priorCTReports>111222</priorCTReports><additionalInfo>Multiple devices used within short time spans</additionalInfo></personOrUserReported>"`;

exports[`buildPersonOrUserReported > should build a person or user reported with only physical details 1`] = `"<personOrUserReported><personOrUserReportedPerson><firstName>Robert</firstName><lastName>Brown</lastName><address><address>456 Oak Street</address><city>Chicago</city><state>IL</state></address><age>42</age></personOrUserReportedPerson><vehicleDescription>Blue Ford F-150</vehicleDescription><estimatedLocation><city>Chicago</city><region>Illinois</region><countryCode>US</countryCode></estimatedLocation></personOrUserReported>"`;
